ledger/test/test_ledger.py

```diff
@@ -215,6 +204,66 @@ def test_consistency_verification_on_startup_case_2(tempdir):
     ledger.stop()
 
 
+def test_recover_merkle_tree_invalid_hash_store(create_ledger_callable, tempdir,
+                                                txn_serializer, hash_serializer, genesis_txn_file):
+    '''
+    Check that tree can be recovered from txn log if recovering from hash store failed
+    (we have one more txn in hash store than in txn log, so consistency verification fails).
+    '''
+    ledger = create_ledger_callable(
+        txn_serializer, hash_serializer, tempdir, genesis_txn_file)
+    for d in range(5):
+        ledger.add(random_txn(d))
+
+    size_before = ledger.size
+    tree_root_hash_before = ledger.tree.root_hash
+    tree_size_before = ledger.tree.tree_size
+    root_hash_before = ledger.root_hash
+    hashes_before = ledger.tree.hashes
+
+    # add to hash store only
+    ledger._addToTree(ledger.serialize_for_tree(random_txn(50)),
+                      serialized=True)
+    ledger.stop()
+
+    restartedLedger = create_ledger_callable(txn_serializer,
+                                             hash_serializer, tempdir, genesis_txn_file)
+
+    assert size_before == restartedLedger.size
+    assert root_hash_before == restartedLedger.root_hash
+    assert hashes_before == restartedLedger.tree.hashes
+    assert tree_root_hash_before == restartedLedger.tree.root_hash
+    assert tree_size_before == restartedLedger.tree.tree_size
+
+
+def test_recover_merkle_tree_invalid_txn_log(create_ledger_callable, tempdir,
```

Comment:
It seems that it would be more accurate to name this test `test_recover_merkle_tree_from_txn_log_if_hash_store_lags_behind` and the previous test `test_recover_merkle_tree_from_txn_log_if_hash_store_runs_ahead`.

Reply:
ok